### PR TITLE
docs(enriching): added docs for enrichment tables

### PR DESCRIPTION
**Pull request by StephenWakely**
This adds the docs for the enrichment tables.

There is a new section for `enrichment_tables` under components. The `file` enrichment table is docemented there.

Docs for 2 VRL functions are included
- `get_enrichment_table_record` 
-  `find_enrichment_table_records`

An example config for this section would look like the following:

Signed-off-by: Stephen Wakely <fungus.humungus@gmail.com>

### CODE REVIEW

File: website/cue/reference/remap/functions/get_enrichment_table_record.cue
Comment (lucperkins) [inline website/cue/reference/remap/functions/get_enrichment_table_record.cue:29]: Missing a comma here

File: website/cue/reference/components.cue
Comment (StephenWakely) [inline website/cue/reference/components.cue:42]: I presume stateful is not a valid concept in enrichment tables. However, not adding the field here causes errors on line 1202 of this file where it attempts to check the value of this field. I'm not sure how to fix those errors.

**Review by spencergilbert — APPROVED** [review]

File: website/cue/reference/components/enrichment_tables/file.cue
Comment (binarylogic) [inline website/cue/reference/components/enrichment_tables/file.cue:1]: This isn't a component and it shouldn't follow the component behavior or patterns. We also use "component" to refer to actual nodes in the processing graph. We should just add a root level file in the [`reference` folder](https://github.com/vectordotdev/vector/tree/master/website/cue/reference) called `enrichment_tables.cue`.

File: website/cue/reference/components/enrichment_tables/file.cue
Comment (binarylogic) [inline website/cue/reference/components/enrichment_tables/file.cue:3]: None of this configuration makes sense for enrichment tables, hence my comment above. Things like `title`, `classes`, `support`, etc just don't make sense in this context. I'd prefer to rework the docs so they are reflective of the actual feature and not wedged in as a component which feels awkward.

**Review by binarylogic — CHANGES_REQUESTED** [review]
Thanks for taking a stab at this! I'd like to see us properly document the root level `enrichment_tables` concept instead of trying to wedge it in as a new component type.

**Review by lucperkins — APPROVED** [review]
There's more we can do here but this gets us into release shape, I think

**Review by jszwedko — APPROVED** [review]
Agreed, thanks for your work on this @lucperkins and @StephenWakely !

File: website/cue/reference/components.cue
Comment (JeanMertz) [inline website/cue/reference/components.cue:142]: When discussing "components" in Vector, I always considered them to be sinks/sources or transforms, and all of them are part of a "pipeline".

Are we sure we want to muddy the waters here by also chucking enrichment tables within that category? I'd personally vote against that decision.

It seems to me the closer analogy to enrichment tables are environment variables: we provide a way to access the data they store, but it's not something you directly hook into a pipeline. Should they be in a new category of "data providers" or something similar?

I realize this is our Cue docs, and not exposed publicly, but from experience, once terminology gets muddied, it seeps through in other places, and before you know it, when discussing "components", half of the people in a meeting think you're talking about enrichment tables, and the other half consider those to be of a different kind, making it more difficult to have precise discussions on the topic.

File: website/cue/reference/configuration.cue
Comment (JeanMertz) [inline website/cue/reference/configuration.cue:51]: I assume this meant to say "Vector needs read-access to the enrichment table file"?

Now it reads as if flipping the write-bit prevents you from using the file as an enrichment table.

File: website/cue/reference/configuration.cue
Comment (JeanMertz) [inline website/cue/reference/configuration.cue:57]: Do we support path expansion? (e.g. `~/foo/bar.csv`), if so, it's worth adding that as an example.

File: website/cue/reference/glossary.cue
Comment (JeanMertz) [inline website/cue/reference/glossary.cue:61]: I think there are plans to potentially provide access to non-file-based information. Although I guess we could update this description if/when that happens.

File: website/cue/reference/remap/functions.cue
Comment (JeanMertz) [inline website/cue/reference/remap/functions.cue:39]: Nit: we might call this `External Data` instead, and chuck  `get_env_var` into this category as well.

File: website/cue/reference/remap/functions/find_enrichment_table_records.cue
Comment (JeanMertz) [inline website/cue/reference/remap/functions/find_enrichment_table_records.cue:43]: This is out-of-scope in terms of code changes, but shouldn't this flag be configurable _per field_ in a search?

That is, I'd expect to write:

```coffee
conditions = {
	"my_field_1": { "value": "my_case_sensitive_value", "case_sensitive": true },
	"my_field_2": { "value": "my_case_insensitive_value", "case_sensitive": false },
}

find_enrichment_table_records("my_table", conditions)
```

As opposed to:

```coffee
conditions = {
	"my_field_1": "my_case_sensitive_value",
	"my_field_2": "my_case_insensitive_value",
}

find_enrichment_table_records("my_table", conditions, case_sensitive: true)
```

The former also allows us to add more options for each condition in the future (e.g. we could add a `match_type` field and have it be either `exact` or `partial` or whatever).

I know this is a very late call for such a change, but this came to mind while reading the Cue docs, and I felt it would be better not to let this opportunity go to waste before we officially release enrichment tables.

File: website/cue/reference/remap/functions/find_enrichment_table_records.cue
Comment (JeanMertz) [inline website/cue/reference/remap/functions/find_enrichment_table_records.cue:60]: Gotta love the Cue formatter 😅

File: website/cue/reference/remap/functions/get_enrichment_table_record.cue
Comment (JeanMertz) [inline website/cue/reference/remap/functions/get_enrichment_table_record.cue:49]: I suspect people will want to differentiate between these two errors. There's no actionable here right now, but I do wonder if at some point we'll want to switch to typed errors in VRL, so that people can pattern-match against them.

File: website/cue/reference/remap/functions/get_enrichment_table_record.cue
Comment (JeanMertz) [inline website/cue/reference/remap/functions/get_enrichment_table_record.cue:52]: Given there are two error states, and each state has its own error message, it might be worth providing an example on how to differentiate between the two errors (e.g. comparing `err` against both strings).

**Review by JeanMertz — CHANGES_REQUESTED** [review]
A few nits, and two comments that I think are worth discussing/solving before we merge.

File: website/cue/reference/configuration.cue
Comment (StephenWakely) [reply website/cue/reference/configuration.cue:57]: I'm not sure about how this would work with Windows?

File: website/cue/reference/components.cue
Comment (JeanMertz) [reply website/cue/reference/components.cue:142]: I just noticed @binarylogic made a similar comment:

> Thanks for taking a stab at this! I'd like to see us properly document the root level enrichment_tables concept instead of trying to wedge it in as a new component type.

Maybe we forgot to remove it from this list?

File: website/cue/reference/remap/functions/find_enrichment_table_records.cue
Comment (StephenWakely) [reply website/cue/reference/remap/functions/find_enrichment_table_records.cue:43]: I think that it is highly unlikely that you would want case sensitive in one field and insensitive in another, and so decided that the extra complexity wasn't worth it.

If a use case comes along we could add the ability to override the global `case_sensitive` on a field basis. I think it is worth holding off until then.

File: website/cue/reference/components.cue
Comment (lucperkins) [reply website/cue/reference/components.cue:142]: @JeanMertz Yep, just forgot to remove it from this list. Enrichment tables are no longer listed as a component type.

File: website/cue/reference/remap/functions/find_enrichment_table_records.cue
Comment (jszwedko) [reply website/cue/reference/remap/functions/find_enrichment_table_records.cue:43]: While I agree with Jean that users may want to search insensitive based on certain fields, this also seems like something we could do in additive fashion later while preserving compatibility so I'm inclined to release as-is to start collecting user feedback. We can schedule more work around this in the future.

File: website/cue/reference/remap/functions/get_enrichment_table_record.cue
Comment (jszwedko) [reply website/cue/reference/remap/functions/get_enrichment_table_record.cue:49]: Yeah, this came up before too (I think in the PR adding the functions? Or maybe the RFC?). I completely agree that users are likely to want to differentiate these errors. I think we will probably have more VRL functions in the future that have discrete errors too. Definitely worth mulling about.

File: website/cue/reference/remap/functions/get_enrichment_table_record.cue
Comment (jszwedko) [reply website/cue/reference/remap/functions/get_enrichment_table_record.cue:52]: I hesitate to push people to use string error matching since it is awkward and brittle (we make no compatibility guarantees about error strings). I think we should generally only provide examples of handling errors opaquely until we provide better utilities for differentiating errors.

File: website/cue/reference/remap/functions.cue
Comment (jszwedko) [reply website/cue/reference/remap/functions.cue:39]: I think I like keeping functions that specifically interact with enrichment tables in their own section, but I don't have strong opinions.

File: website/cue/reference/configuration.cue
Comment (jszwedko) [reply website/cue/reference/configuration.cue:57]: Windows users also wouldn't typically be able to use `/data/info.csv` either (it'd be something like `C:\data\info.csv`.

If we wanted to limit it to cross-platform examples, which I think is reasonable, we'd be limited to relative paths.

This is a small nit though, I think it's fine to merge as-is with or without an example using `~`.

**Comment by jszwedko** [comment]
@JeanMertz tagging you again for any final thoughts before this is merged. I think everything was addressed or responded to.

File: website/cue/reference/remap/functions/get_enrichment_table_record.cue
Comment (JeanMertz) [reply website/cue/reference/remap/functions/get_enrichment_table_record.cue:52]: That makes sense to me.

File: website/cue/reference/remap/functions/find_enrichment_table_records.cue
Comment (JeanMertz) [reply website/cue/reference/remap/functions/find_enrichment_table_records.cue:43]: Hmm, yeah, we might be able to do this without breaking backward compatibility (or accept a breakage if needed). 👍

File: website/cue/reference/remap/functions.cue
Comment (JeanMertz) [reply website/cue/reference/remap/functions.cue:39]: That is a fair point. I didn't think we'd get any other than the two we have now, but that might not be the case. We can go with the current set-up, and see if changes are needed in the future.

**Review by JeanMertz — APPROVED** [review]

**Pull request merged**
